rt/frontend/jaxrs/src/main/java/org/apache/cxf/jaxrs/utils/FormUtils.java

```diff
@@ -175,8 +176,20 @@ public static void populateMapFromString(MultivaluedMap<String, String> params,
             && MessageUtils.getContextualBoolean(m, FORM_PARAMS_FROM_HTTP_PARAMS, true)) {
             for (Enumeration<String> en = request.getParameterNames(); en.hasMoreElements();) {
                 String paramName = en.nextElement();
-                String[] values = request.getParameterValues(paramName);
-                params.put(HttpUtils.urlDecode(paramName), Arrays.asList(values));
+                String[] parameterValues = request.getParameterValues(paramName);
+
+                // these parameters will already be URLdecoded by the servlet container on the
+                // request.getParameterValues() call above
+
+                List<String> values = Arrays.asList(parameterValues);
+
+                if (!decode) {
```

Comment:
I think we should not encode the input. I understand that you are trying to mitigate the decoding issue(s), I am looking into the issue, will try to find out alternative solution.

Reply:
I'm not a fan of the solution, which is probably a slightly bizarre comment to make on my own code. There is a scenario where CXF is reading parameters from HttpServetReques.getParameter(), where the parameter itself is annotated with `@Encoded`. In that scenario, I'm not sure how we can provide the parameter correctly without re-encoding.

Reply:
not using getParameter but the body inputstream - think cxf already has some part of it in body readers? - can solves it.

another side note is that moving the evaluation to be lazy instead of eager in a list can also be a good thing and avoid a lot of preprocessing (think `if (param1 is ok) { processBigParam(param2); }`, if param1 is not ok you don't care of having processed param2 and worse if you have some security checks in your endpoint (`@Context HttpServletRequest` is not that rare ;)). This can be done in a custom MultivaluedMap impl (new RequestParametersMultivaluedMap<>(request)).

Hope it makes sense.

Reply:
> not using getParameter but the body inputstream - think cxf already has some part of it in body readers? - can solves it.

Yes, the  body inputstream is handled right before